File: app/src/main/kotlin/com/gtaclone/world/City3D.kt
package com.gtaclone.world

import kotlin.random.Random

data class Box3D(
    val cx: Float, val cy: Float, val cz: Float,
    val w: Float,  val h: Float,  val d: Float,
    val rotY: Float = 0f,
    val r: Float,  val g: Float,  val b: Float
)

class City3D {
    val BLOCK = 40f
    val ROAD  = 10f
    val GRID  = 10

    val buildings  = mutableListOf<Box3D>()
    val roadSlabs  = mutableListOf<Box3D>()
    val parkTrees  = mutableListOf<Box3D>()

    private val colliders = mutableListOf<FloatArray>()  // [minX, minZ, maxX, maxZ]

    init { generate() }

    private fun bs(n: Int) = n * (BLOCK + ROAD)  // block start

    private fun generate() {
        val rng = Random(42)

        // Road slabs (slightly raised flat quads)
        val total = (GRID * (BLOCK + ROAD)).toFloat()
        for (n in 0 until GRID - 1) {
            val roadCenter = bs(n) + BLOCK + ROAD / 2
            // Horizontal road (runs along X)
            roadSlabs.add(Box3D(total / 2, 0.05f, roadCenter, total, 0.1f, ROAD, 0f, 0.27f, 0.27f, 0.27f))
            // Vertical road (runs along Z)
            roadSlabs.add(Box3D(roadCenter, 0.05f, total / 2, ROAD, 0.1f, total, 0f, 0.27f, 0.27f, 0.27f))
        }

        // Buildings
        val palette = listOf(
            Triple(0.48f, 0.38f, 0.22f), Triple(0.30f, 0.44f, 0.55f),
            Triple(0.55f, 0.30f, 0.30f), Triple(0.42f, 0.42f, 0.44f),
            Triple(0.35f, 0.50f, 0.34f)
        )

        for (bx in 0 until GRID) {
            for (bz in 0 until GRID) {
                // Park at blocks 4-5
                if (bx in 4..5 && bz in 4..5) { spawnPark(bs(bx), bs(bz)); continue }

                val x0 = bs(bx).toFloat(); val z0 = bs(bz).toFloat()
                val num = rng.nextInt(2) + 1
                for (i in 0 until num) {
                    val margin = 4f; val avail = BLOCK - margin * 2
                    val bw = avail * (0.35f + rng.nextFloat() * 0.55f)
                    val bd = avail * (0.35f + rng.nextFloat() * 0.55f)
                    val bh = 5f + rng.nextFloat() * 38f
                    val ox = margin + rng.nextFloat() * (avail - bw)
                    val oz = margin + rng.nextFloat() * (avail - bd)
                    val cx = x0 + ox + bw / 2; val cz = z0 + oz + bd / 2
                    val (r, g, b) = palette[(bx + bz + i) % palette.size]
                    buildings.add(Box3D(cx, bh / 2, cz, bw, bh, bd, 0f, r, g, b))
                    colliders.add(floatArrayOf(x0 + ox, z0 + oz, x0 + ox + bw, z0 + oz + bd))
                }
            }
        }
    }

    private fun spawnPark(x0: Float, z0: Float) {
        // Tree trunks + canopies
        val treePos = listOf(8f to 8f, 30f to 8f, 8f to 30f, 30f to 30f, 19f to 19f)
        treePos.forEach { (tx, tz) ->
            parkTrees.add(Box3D(x0 + tx, 1.5f, z0 + tz, 1f, 3f, 1f, 0f, 0.36f, 0.22f, 0.10f))
            parkTrees.add(Box3D(x0 + tx, 5f,   z0 + tz, 5f, 4f, 5f, 0f, 0.20f, 0.60f, 0.20f))
        }
    }

    fun isBlocked(x: Float, z: Float, radius: Float = 0.8f): Boolean =
        colliders.any { a -> x + radius > a[0] && x - radius < a[2] && z + radius > a[1] && z - radius < a[3] }

    fun citySize() = (GRID * (BLOCK + ROAD)).toFloat()
}

File: GTAClone-iOS/GTAClone/world/City3D.swift
import SceneKit

// Simple deterministic LCG so the city looks the same every run
private class RNG {
    private var s: UInt64 = 42
    func next() -> Float {
        s = s &* 6364136223846793005 &+ 1442695040888963407
        return Float((s >> 33) & 0xFFFF) / 65535.0
    }
    func range(_ lo: Float, _ hi: Float) -> Float { lo + next() * (hi - lo) }
    func int(_ n: Int) -> Int { Int(next() * Float(n)) }
}

struct Collider { let minX, minZ, maxX, maxZ: Float }

class City3D {

    let BLOCK: Float = 40; let ROAD: Float = 10; let GRID = 10
    private(set) var colliders: [Collider] = []

    func build(into scene: SCNScene) {
        let rng = RNG()
        let total = Float(GRID) * (BLOCK + ROAD)

        // Ground (SCNFloor = infinite reflective plane at y=0)
        let floor = SCNFloor()
        floor.reflectivity = 0
        floor.firstMaterial?.diffuse.contents = UIColor(r: 0.24, g: 0.42, b: 0.16)
        scene.rootNode.addChildNode(SCNNode(geometry: floor))

        // Road grid (flat slabs, y = 0.02)
        let roadMat = SCNMaterial(); roadMat.diffuse.contents = UIColor(r: 0.27, g: 0.27, b: 0.27)
        for n in 0..<(GRID - 1) {
            let rc = Float(n) * (BLOCK + ROAD) + BLOCK + ROAD / 2
            addSlab(scene: scene, cx: total/2, cz: rc, w: total, d: ROAD, mat: roadMat)
            addSlab(scene: scene, cx: rc, cz: total/2, w: ROAD, d: total, mat: roadMat)
        }

        // Yellow centre-line stripes (very thin, raised slightly above road)
        let lineMat = SCNMaterial(); lineMat.diffuse.contents = UIColor(r: 0.9, g: 0.85, b: 0.1)
        for n in 0..<(GRID - 1) {
            let rc = Float(n) * (BLOCK + ROAD) + BLOCK + ROAD / 2
            addSlab(scene: scene, cx: total/2, cz: rc, w: total, d: 0.25, mat: lineMat, y: 0.06)
            addSlab(scene: scene, cx: rc, cz: total/2, w: 0.25, d: total, mat: lineMat, y: 0.06)
        }

        // Buildings
        let palette: [UIColor] = [
            UIColor(r:0.48,g:0.38,b:0.22), UIColor(r:0.30,g:0.44,b:0.55),
            UIColor(r:0.55,g:0.30,b:0.30), UIColor(r:0.42,g:0.42,b:0.44),
            UIColor(r:0.35,g:0.50,b:0.34)
        ]
        for bx in 0..<GRID {
            for bz in 0..<GRID {
                let x0 = Float(bx) * (BLOCK + ROAD); let z0 = Float(bz) * (BLOCK + ROAD)
                if bx >= 4 && bx <= 5 && bz >= 4 && bz <= 5 { addPark(scene: scene, x0: x0, z0: z0); continue }

                let margin: Float = 4; let avail = BLOCK - margin * 2
                let bw = rng.range(avail * 0.35, avail * 0.90)
                let bd = rng.range(avail * 0.35, avail * 0.90)
                let bh = rng.range(5, 43)
                let ox = margin + rng.range(0, avail - bw)
                let oz = margin + rng.range(0, avail - bd)
                let col = palette[(bx + bz + rng.int(3)) % palette.count]
                addBuilding(scene: scene, cx: x0+ox+bw/2, h: bh, cz: z0+oz+bd/2, w: bw, d: bd, color: col)
                colliders.append(Collider(minX: x0+ox, minZ: z0+oz, maxX: x0+ox+bw, maxZ: z0+oz+bd))
            }
        }
    }

    private func addSlab(scene: SCNScene, cx: Float, cz: Float, w: Float, d: Float, mat: SCNMaterial, y: Float = 0.02) {
        let geo = SCNBox(width: CGFloat(w), height: 0.04, length: CGFloat(d), chamferRadius: 0)
        geo.firstMaterial = mat
        let n = SCNNode(geometry: geo); n.position = SCNVector3(cx, y, cz)
        scene.rootNode.addChildNode(n)
    }

    private func addBuilding(scene: SCNScene, cx: Float, h: Float, cz: Float, w: Float, d: Float, color: UIColor) {
        let geo = SCNBox(width: CGFloat(w), height: CGFloat(h), length: CGFloat(d), chamferRadius: 0.12)
        geo.firstMaterial?.diffuse.contents = color
        let n = SCNNode(geometry: geo); n.position = SCNVector3(cx, h/2, cz)
        scene.rootNode.addChildNode(n)

        // Window rows every 3 m
        let winMat = SCNMaterial(); winMat.diffuse.contents = UIColor(r: 0.85, g: 0.92, b: 0.72)
        winMat.emission.contents = UIColor(r: 0.3, g: 0.35, b: 0.1)
        var floor: Float = 3
        while floor < h - 1 {
            let wy = -(h/2) + floor
            let fGeo = SCNBox(width: CGFloat(w*0.62), height: 0.8, length: 0.06, chamferRadius: 0)
            fGeo.firstMaterial = winMat
            let fn = SCNNode(geometry: fGeo); fn.position = SCNVector3(0, wy, d/2 + 0.04)
            n.addChildNode(fn)
            floor += 3
        }
    }

    private func addPark(scene: SCNScene, x0: Float, z0: Float) {
        let treePos: [(Float, Float)] = [(8,8),(30,8),(8,30),(30,30),(19,19)]
        for (tx, tz) in treePos {
            let trunk = SCNBox(width: 1, height: 3, length: 1, chamferRadius: 0.1)
            trunk.firstMaterial?.diffuse.contents = UIColor(r: 0.36, g: 0.22, b: 0.10)
            let tn = SCNNode(geometry: trunk); tn.position = SCNVector3(x0+tx, 1.5, z0+tz)
            scene.rootNode.addChildNode(tn)
            let canopy = SCNSphere(radius: 3)
            canopy.firstMaterial?.diffuse.contents = UIColor(r: 0.20, g: 0.60, b: 0.20)
            let cn = SCNNode(geometry: canopy); cn.position = SCNVector3(x0+tx, 5.5, z0+tz)
            scene.rootNode.addChildNode(cn)
        }
    }

    func isBlocked(x: Float, z: Float, radius: Float = 0.9) -> Bool {
        colliders.contains { c in x+radius > c.minX && x-radius < c.maxX && z+radius > c.minZ && z-radius < c.maxZ }
    }

    func citySize() -> Float { Float(GRID) * (BLOCK + ROAD) }
}

private extension UIColor {
    convenience init(r: CGFloat, g: CGFloat, b: CGFloat) { self.init(red: r, green: g, blue: b, alpha: 1) }
}
